src/stan_math_backend/Cpp.ml

```diff
@@ -0,0 +1,732 @@
+(** A set of data types representing the C++ we generate *)
+
+open Core_kernel
+
+type identifier = string [@@deriving sexp]
+
+(** C++ types *)
+type type_ =
+  | Auto
+  | Void
+  | Int
+  | Double
+  | Complex of type_
+  | TemplateType of identifier
+  | Vector of type_
+  | Array of type_ * int
+  | Type_literal of identifier  (** Used for things like Eigen::Index *)
+  | Matrix of type_ * int * int
+  | Ref of type_
+  | Const of type_
+  | Pointer of type_
+  | TypeTrait of string * type_ list
+      (** e.g. stan::promote_scalar, stan:base_type *)
+[@@deriving sexp]
+
+module Types = struct
+  (** Helpers for constructing types *)
+
+  let local_scalar = Type_literal "local_scalar_t__"
+
+  (** A [std::vector<t>] *)
+  let std_vector t = Vector t
+
+  let bool = Type_literal "bool"
+  let complex s = Complex s
+
+  (** An [Eigen::Matrix<s, -1, 1>]*)
+  let vector s = Matrix (s, -1, 1)
+
+  (** An [Eigen::Matrix<s, 1, -1>]*)
+  let row_vector s = Matrix (s, 1, -1)
+
+  (** An [Eigen::Matrix<s, -1, -1>]*)
+  let matrix s = Matrix (s, -1, -1)
+
+  (** A [std::string]*)
+  let string = Type_literal "std::string"
+
+  let size_t = Type_literal "size_t"
+  let const_ref t = Const (Ref t)
+  let const_char_array i = Array (Const (Pointer (Type_literal "char")), i)
+end
+
+type operator =
+  | Multiply
+  | Divide
+  | Add
+  | Subtract
+  | Eq
+  | LEq
+  | GEq
+  | Lthn
+  | Gthn
+  | And
+  | Or
+[@@deriving sexp]
+
+type unary_op = PMinus | Incr [@@deriving sexp]
+
+type expr =
+  | Literal of string  (** printed as-is *)
+  | Var of identifier
+  | Parens of expr
+  | FunCall of identifier * type_ list * expr list
+  | MethodCall of expr * identifier * type_ list * expr list
+  | StaticMethodCall of type_ * identifier * type_ list * expr list
+  | Constructor of type_ * expr list  (** printed as [type(expr1, expr2, ...)]*)
+  | InitializerExpr of type_ * expr list
+      (** printed as [type{expr1, expr2, ...}]*)
+  | ArrayLiteral of expr list
+  | TernaryIf of expr * expr * expr
+  | Cast of type_ * expr
+  | Index of expr * expr
+  | New of string option * type_ * expr list
+  | Assign of expr * expr  (** NB: Not all exprs are valid lvalues! *)
+  | Unary of unary_op * expr
+  | StreamInsertion of expr * expr list  (** Corresponds to [operator<<] *)
+  | BinOp of expr * operator * expr
+[@@deriving sexp]
+
+module Exprs = struct
+  (** Some helper values and functions *)
+
+  (** Call a method on object, wrapping it in parentheses if
+      it is not a variable
+  *)
+  let method_call obj name templates args =
+    match obj with
+    | Var _ -> MethodCall (obj, name, templates, args)
+    | _ -> MethodCall (Parens obj, name, templates, args)
```

Comment:
This helper method is needed to make sure that the precendence you specify in OCaml is actually what gets printed.

Consider the code example:

```ocaml
let open Expression_syntax in 
let a = Var "a" in
let b = Var "b" in
(a + b).@!("eval")
```

If _not_ for the above logic, this would be a `MethodCall(BinOp(...), ...)`, but that would print as
`a + b.eval()`, which is clearly not the correct order. What we want is `(a + b).eval()`.

The reason we match on a Var is so avoid printing the () in redundant cases like `c.eval()` becoming `(c).eval()`